Mathlib/MeasureTheory/MeasurableSpace/CountablyGenerated.lean

```diff
@@ -99,6 +117,140 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
     CountablyGenerated (α × β) :=
   .sup (.comap Prod.fst) (.comap Prod.snd)
 
+section SeparatesPoints
+
+/-- We say that a measurable space separates points if for any two distinct points,
+there is a measurable set containing one but not the other. -/
+class SeparatesPoints (α : Type*) [m : MeasurableSpace α] : Prop where
+  separates : ∀ x y : α, (∀ s, MeasurableSet s → (x ∈ s → y ∈ s)) → x = y
+
+theorem separatesPoints_def [MeasurableSpace α] [hs : SeparatesPoints α] {x y : α}
+    (h : (∀ s, MeasurableSet s → (x ∈ s → y ∈ s))) : x = y := hs.separates _ _ h
+
+theorem exists_measurableSet_of_neq [MeasurableSpace α] [SeparatesPoints α] {x y : α}
+    (h : x ≠ y) : ∃ s, MeasurableSet s ∧ x ∈ s ∧ y ∉ s := by
+  contrapose! h
+  exact separatesPoints_def h
+
+/-- If the measurable space generated by `S` separates points,
+then this is witnessed by sets in `S`. -/
+theorem separating_of_generateFrom (S : Set (Set α))
+    [h : @SeparatesPoints α (generateFrom S)] :
+    ∀ x y : α, (∀ s ∈ S, x ∈ s ↔ y ∈ s) → x = y := by
+  letI := generateFrom S
+  intros x y hxy
+  rw [← forall_generateFrom_mem_iff_mem_iff] at hxy
+  exact separatesPoints_def $ fun _ hs => (hxy _ hs).mp
+
+theorem SeparatesPoints.mono {m m' : MeasurableSpace α} [hsep : @SeparatesPoints _ m] (h : m ≤ m') :
+    @SeparatesPoints _ m' := @SeparatesPoints.mk _ m' fun _ _ hxy ↦
+    @SeparatesPoints.separates _ m hsep _ _ fun _ hs ↦ hxy _ (h _ hs)
+
+instance (priority := 100) Subtype.separatesPoints [MeasurableSpace α] [h : SeparatesPoints α]
+    {s : Set α} : SeparatesPoints s :=
+  ⟨fun _ _ hxy ↦ Subtype.val_injective $ h.1 _ _ fun _ ht ↦ hxy _ $ measurable_subtype_coe ht⟩
+
+instance (priority := 100) separatesPoints_of_measurableSingletonClass [MeasurableSpace α]
+    [MeasurableSingletonClass α] : SeparatesPoints α := by
+  refine' ⟨fun x y h ↦ _⟩
```

Comment:
```suggestion
  refine ⟨fun x y h ↦ ?_⟩
```
refine' was used by mathport to stay close to lean3 syntax, but refine is preferred now.